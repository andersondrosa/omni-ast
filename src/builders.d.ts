import { Expression, Identifier, MemberExpression } from "./types";

declare function memberExpression(
  object: Expression,
  property: Expression | Identifier,
  computed?: boolean,
  optional?: true | false | null
): MemberExpression;
