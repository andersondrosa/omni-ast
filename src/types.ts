export type JsonTypes = string | number | boolean | null | object | [];

export interface BaseNodeWithoutComments {
  type: string;
  loc?: SourceLocation | null | undefined;
  range?: [number, number] | undefined;
}

export interface BaseNode extends BaseNodeWithoutComments {
  leadingComments?: Comment[] | undefined;
  trailingComments?: Comment[] | undefined;
}

export interface NodeMap {
  JsonExpression: JsonExpression;
  AssignmentProperty: AssignmentProperty;
  CatchClause: CatchClause;
  Expression: Expression;
  Function: Function;
  Identifier: Identifier;
  Literal: Literal;
  Pattern: Pattern;
  Program: Program;
  Property: Property;
  PropertyDefinition: PropertyDefinition;
  SpreadElement: SpreadElement;
  Statement: Statement;
  SwitchCase: SwitchCase;
  VariableDeclarator: VariableDeclarator;
}

export type Node = NodeMap[keyof NodeMap];

export interface SourceLocation {
  source?: string | null | undefined;
  start: Position;
  end: Position;
}

export interface Position {
  line: number;
  column: number;
}

export interface Program extends BaseNode {
  type: "Program";
  sourceType: "script" | "module";
  body: Array<Directive | Statement>;
  comments?: Comment[] | undefined;
}

export interface Directive extends BaseNode {
  type: "ExpressionStatement";
  expression: Literal;
  directive: string;
}

export interface BaseFunction extends BaseNode {
  params: Pattern[];
  generator?: boolean | undefined;
  async?: boolean | undefined;
  body: BlockStatement | Expression;
}

export type Function =
  | FunctionDeclaration
  | FunctionExpression
  | ArrowFunctionExpression;

export type Statement =
  | ExpressionStatement
  | BlockStatement
  | StaticBlock
  | EmptyStatement
  | WithStatement
  | ReturnStatement
  | BreakStatement
  | ContinueStatement
  | IfStatement
  | SwitchStatement
  | ThrowStatement
  | TryStatement
  | WhileStatement
  | DoWhileStatement
  | ForStatement
  | ForInStatement
  | ForOfStatement
  | Declaration;

export type UnaryOperator =
  | "-"
  | "+"
  | "!"
  | "~"
  | "typeof"
  | "void"
  | "delete";

export type BinaryOperator =
  | "=="
  | "!="
  | "==="
  | "!=="
  | "<"
  | "<="
  | ">"
  | ">="
  | "<<"
  | ">>"
  | ">>>"
  | "+"
  | "-"
  | "*"
  | "/"
  | "%"
  | "**"
  | "|"
  | "^"
  | "&"
  | "in"
  | "instanceof";

export type LogicalOperator = "||" | "&&" | "??";

export type AssignmentOperator =
  | "="
  | "+="
  | "-="
  | "*="
  | "/="
  | "%="
  | "**="
  | "<<="
  | ">>="
  | ">>>="
  | "|="
  | "^="
  | "&="
  | "||="
  | "&&="
  | "??=";

export type UpdateOperator = "++" | "--";

export interface BaseStatement extends BaseNode {}

export interface EmptyStatement extends BaseStatement {
  type: "EmptyStatement";
}

export interface BlockStatement extends BaseStatement {
  type: "BlockStatement";
  body: Statement[];
  innerComments?: Comment[] | undefined;
}

export interface StaticBlock extends Omit<BlockStatement, "type"> {
  type: "StaticBlock";
}

export interface ExpressionStatement extends BaseStatement {
  type: "ExpressionStatement";
  expression: Expression;
}

export interface IfStatement extends BaseStatement {
  type: "IfStatement";
  test: Expression;
  consequent: Statement;
  alternate?: Statement | null | undefined;
}

export interface LabeledStatement extends BaseStatement {
  type: "LabeledStatement";
  label: Identifier;
  body: Statement;
}

export interface BreakStatement extends BaseStatement {
  type: "BreakStatement";
  label?: Identifier | null | undefined;
}

export interface ContinueStatement extends BaseStatement {
  type: "ContinueStatement";
  label?: Identifier | null | undefined;
}

export interface WithStatement extends BaseStatement {
  type: "WithStatement";
  object: Expression;
  body: Statement;
}

export interface SwitchStatement extends BaseStatement {
  type: "SwitchStatement";
  discriminant: Expression;
  cases: SwitchCase[];
}

export interface ReturnStatement extends BaseStatement {
  type: "ReturnStatement";
  argument?: Expression | null | undefined;
}

export interface ThrowStatement extends BaseStatement {
  type: "ThrowStatement";
  argument: Expression;
}

export interface TryStatement extends BaseStatement {
  type: "TryStatement";
  block: BlockStatement;
  handler?: CatchClause | null | undefined;
  finalizer?: BlockStatement | null | undefined;
}

export interface WhileStatement extends BaseStatement {
  type: "WhileStatement";
  test: Expression;
  body: Statement;
}

export interface DoWhileStatement extends BaseStatement {
  type: "DoWhileStatement";
  body: Statement;
  test: Expression;
}

export interface ForStatement extends BaseStatement {
  type: "ForStatement";
  init?: VariableDeclaration | Expression | null | undefined;
  test?: Expression | null | undefined;
  update?: Expression | null | undefined;
  body: Statement;
}

export interface BaseForXStatement extends BaseStatement {
  left: VariableDeclaration | Pattern;
  right: Expression;
  body: Statement;
}

export interface ForInStatement extends BaseForXStatement {
  type: "ForInStatement";
}

export interface DebuggerStatement extends BaseStatement {
  type: "DebuggerStatement";
}

export type Declaration = FunctionDeclaration | VariableDeclaration;

export interface BaseDeclaration extends BaseStatement {}

export interface MaybeNamedFunctionDeclaration
  extends BaseFunction,
    BaseDeclaration {
  type: "FunctionDeclaration";
  id: Identifier | null;
  body: BlockStatement;
}

export interface FunctionDeclaration extends MaybeNamedFunctionDeclaration {
  id: Identifier;
}

export interface VariableDeclaration extends BaseDeclaration {
  type: "VariableDeclaration";
  declarations: VariableDeclarator[];
  kind: "var" | "let" | "const";
}

export interface VariableDeclarator extends BaseNode {
  type: "VariableDeclarator";
  id: Pattern;
  init?: Expression | null | undefined;
}

export interface ExpressionMap {
  ArrayExpression: ArrayExpression;
  ArrowFunctionExpression: ArrowFunctionExpression;
  AssignmentExpression: AssignmentExpression;
  AwaitExpression: AwaitExpression;
  BinaryExpression: BinaryExpression;
  CallExpression: CallExpression;
  ChainExpression: ChainExpression;
  ConditionalExpression: ConditionalExpression;
  FunctionExpression: FunctionExpression;
  Identifier: Identifier;
  Literal: Literal;
  LogicalExpression: LogicalExpression;
  MemberExpression: MemberExpression;
  NewExpression: NewExpression;
  ObjectExpression: ObjectExpression;
  SequenceExpression: SequenceExpression;
  TaggedTemplateExpression: TaggedTemplateExpression;
  TemplateLiteral: TemplateLiteral;
  ThisExpression: ThisExpression;
  UnaryExpression: UnaryExpression;
  UpdateExpression: UpdateExpression;
}

export type Expression = ExpressionMap[keyof ExpressionMap];

export interface BaseExpression extends BaseNode {}

export type ChainElement = SimpleCallExpression | MemberExpression;

export interface ChainExpression extends BaseExpression {
  type: "ChainExpression";
  expression: ChainElement;
}

export interface ThisExpression extends BaseExpression {
  type: "ThisExpression";
}

export interface ArrayExpression extends BaseExpression {
  type: "ArrayExpression";
  elements: Array<Expression | null>; // SpreadElement
}

export interface ObjectExpression extends BaseExpression {
  type: "ObjectExpression";
  properties: Property[]; // SpreadElement
}

export interface Property extends BaseNode {
  type: "Property";
  key: Expression;
  value: Expression | Pattern; // Could be an AssignmentProperty
  kind: "init" | "get" | "set";
  method: boolean;
  shorthand: boolean;
  computed: boolean;
}

export interface PropertyDefinition extends BaseNode {
  type: "PropertyDefinition";
  key: Expression;
  value?: Expression | null | undefined;
  computed: boolean;
  static: boolean;
}

export interface FunctionExpression extends BaseFunction, BaseExpression {
  id?: Identifier | null | undefined;
  type: "FunctionExpression";
  body: BlockStatement;
}

export interface SequenceExpression extends BaseExpression {
  type: "SequenceExpression";
  expressions: Expression[];
}

export interface UnaryExpression extends BaseExpression {
  type: "UnaryExpression";
  operator: UnaryOperator;
  prefix: true;
  argument: Expression;
}

export interface BinaryExpression extends BaseExpression {
  type: "BinaryExpression";
  operator: BinaryOperator;
  left: Expression;
  right: Expression;
}

export interface AssignmentExpression extends BaseExpression {
  type: "AssignmentExpression";
  operator: AssignmentOperator;
  left: Pattern | MemberExpression;
  right: Expression;
}

export interface UpdateExpression extends BaseExpression {
  type: "UpdateExpression";
  operator: UpdateOperator;
  argument: Expression;
  prefix: boolean;
}

export interface LogicalExpression extends BaseExpression {
  type: "LogicalExpression";
  operator: LogicalOperator;
  left: Expression;
  right: Expression;
}

export interface ConditionalExpression extends BaseExpression {
  type: "ConditionalExpression";
  test: Expression;
  alternate: Expression;
  consequent: Expression;
}

export interface BaseCallExpression extends BaseExpression {
  callee: Expression;
  arguments: Array<Expression | SpreadElement>;
}
export type CallExpression = SimpleCallExpression | NewExpression;

export interface SimpleCallExpression extends BaseCallExpression {
  type: "CallExpression";
  optional: boolean;
}

export interface NewExpression extends BaseCallExpression {
  type: "NewExpression";
}

export interface MemberExpression extends BaseExpression, BasePattern {
  type: "MemberExpression";
  object: Expression;
  property: Expression;
  computed: boolean;
  optional: boolean;
}

export type Pattern =
  | Identifier
  | RestElement
  | AssignmentPattern
  | MemberExpression
  | ObjectPattern
  | JsonExpression;

export interface BasePattern extends BaseNode {}

export interface SwitchCase extends BaseNode {
  type: "SwitchCase";
  test?: Expression | null | undefined;
  consequent: Statement[];
}

export interface CatchClause extends BaseNode {
  type: "CatchClause";
  param: Pattern | null;
  body: BlockStatement;
}

export interface JsonExpression extends BaseNode, BaseExpression, BasePattern {
  type: "JsonExpression";
  body: JsonTypes;
}

export interface Identifier extends BaseNode, BaseExpression, BasePattern {
  type: "Identifier";
  name: string;
}

export type Literal = SimpleLiteral | RegExpLiteral | BigIntLiteral;

export interface SimpleLiteral extends BaseNode, BaseExpression {
  type: "Literal";
  value: string | boolean | number | null;
  raw?: string | undefined;
}

export interface RegExpLiteral extends BaseNode, BaseExpression {
  type: "Literal";
  value?: RegExp | null | undefined;
  regex: {
    pattern: string;
    flags: string;
  };
  raw?: string | undefined;
}

export interface BigIntLiteral extends BaseNode, BaseExpression {
  type: "Literal";
  value?: bigint | null | undefined;
  bigint: string;
  raw?: string | undefined;
}
export interface ForOfStatement extends BaseForXStatement {
  type: "ForOfStatement";
  await: boolean;
}

export interface SpreadElement extends BaseNode {
  type: "SpreadElement";
  argument: Expression;
}

export interface ArrowFunctionExpression extends BaseExpression, BaseFunction {
  type: "ArrowFunctionExpression";
  expression: boolean;
  body: BlockStatement | Expression;
}

export interface TemplateLiteral extends BaseExpression {
  type: "TemplateLiteral";
  quasis: TemplateElement[];
  expressions: Expression[];
}

export interface TaggedTemplateExpression extends BaseExpression {
  type: "TaggedTemplateExpression";
  tag: Expression;
  quasi: TemplateLiteral;
}

export interface TemplateElement extends BaseNode {
  type: "TemplateElement";
  tail: boolean;
  value: {
    cooked?: string | null | undefined;
    raw: string;
  };
}

export interface AssignmentProperty extends Property {
  value: Pattern;
  kind: "init";
  method: boolean;
}

export interface ObjectPattern extends BasePattern {
  type: "ObjectPattern";
  properties: Array<AssignmentProperty | RestElement>;
}

export interface RestElement extends BasePattern {
  type: "RestElement";
  argument: Pattern;
}

export interface AssignmentPattern extends BasePattern {
  type: "AssignmentPattern";
  left: Pattern;
  right: Expression;
}

export interface AwaitExpression extends BaseExpression {
  type: "AwaitExpression";
  argument: Expression;
}
