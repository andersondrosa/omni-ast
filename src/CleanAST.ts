//
const remove = ["start", "end", "generator"];

const hideIfNil = [
  "async",
  "id",
  "label",
  "computed",
  "optional",
  "shorthand",
  "prefix",
  "method",
  "expression",
];

export function cleanAST(ast): any {
  //
  if (typeof ast != "object") throw Error("Invalid AST");

  if (ast === null) return null;

  const res = {};

  if (Array.isArray(ast)) return ast.map(cleanAST);

  Object.entries(ast).forEach(([key, value]) => {
    //
    if (remove.includes(key)) return;

    if (hideIfNil.includes(key) && !value) return;

    if (typeof value != "object" || value instanceof RegExp || value === null)
      return (res[key] = value);

    res[key] = cleanAST(value);
  });

  return res;
}
