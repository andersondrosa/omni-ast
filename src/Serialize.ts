import * as Types from "./types";
import { stringify } from "./JsonGenerate";
import { BaseNode } from "./types";

// (data: JsonTypes | BaseNode) {
//   if (!data || typeof data != "object") return JSON.stringify(data);

//   if (Array.isArray(data)) return data.map((x) => generate(x)).join("; ");

//   if (data.hasOwnProperty("type") && (data as any).type) {
//     const node = data as BaseNode;
//     if (nodes.hasOwnProperty(node.type)) return nodes[node.type](node);
//     throw Error("Unknown type: " + node.type);
//   }
// }

export const serialize = (node: BaseNode) => {
  if (!node) return "";
  if (!node.type) return JSON.stringify(node);
  if (nodes.hasOwnProperty(node.type)) return nodes[node.type](node);
  throw Error("Unknown type: " + node.type);
};

export const generate = serialize;

const comma = ", ";

export const JsonExpression = (node) => {
  return stringify(node.body);
};

export const Identifier = (node: Types.Identifier) => {
  return node.name;
};

export const Literal = (node) => {
  return typeof node.value == "string" ? node.raw : String(node.value);
};

export const BlockStatement = (node: Types.BlockStatement) => {
  const statements = node.body.map(serialize).join("; ");
  return `{ ${statements} }`;
};

export const ExpressionStatement = (node: Types.ExpressionStatement) => {
  return serialize(node.expression);
};

export const MemberExpression = (node: Types.MemberExpression) => {
  return serialize(node.object) + "." + serialize(node.property);
};

export const CallExpression = (node: Types.CallExpression) => {
  return `${serialize(node.callee)}(${node.arguments
    .map(serialize)
    .join(comma)})`;
};

export const ArrayExpression = (node: Types.ArrayExpression) => {
  return `[${node.elements.map(serialize).join(comma)}]`;
};

export const ArrowFunctionExpression = (
  node: Types.ArrowFunctionExpression
) => {
  const params = node.params.map(Identifier as any).join(comma);
  return `(${params}) => ${serialize(node.body)}`;
};

export const FunctionExpression = (node: Types.FunctionExpression) => {
  const params = node.params.map(Identifier as any).join(comma);
  return `function ${serialize(node.id)}(${params}) ${serialize(node.body)}`;
};

export const VariableDeclaration = (stmt: Types.VariableDeclaration) => {
  return `${stmt.kind} ${stmt.declarations.map(VariableDeclarator).join(", ")}`;
};

export const VariableDeclarator = (stmt: Types.VariableDeclarator) => {
  if (!stmt.init) return serialize(stmt.id);
  return `${serialize(stmt.id)} = ${serialize(stmt.init)}`;
};

export const ReturnStatement = (stmt: Types.ReturnStatement) => {
  return "return " + serialize(stmt.argument);
};

export const IfStatement = (stmt: Types.IfStatement) => {
  let code = `if (${serialize(stmt.test)}) ${serialize(stmt.consequent)}`;
  if (stmt.alternate) {
    code += ` else ${serialize(stmt.alternate)}`;
  }
  return code;
};

export const BreakStatement = () => {
  return "break";
};

export const ContinueStatement = (node) => {
  return "continue";
};

export const ThrowStatement = (node: Types.ThrowStatement) => {
  return "throw " + serialize(node.argument);
};

export const TryStatement = (node: Types.TryStatement) => {
  let code = `try { ${serialize(node.block)} }`;
  if (node.handler) {
    code += `catch (${serialize(node.handler.param)}) { ${serialize(
      node.handler.body
    )} }`;
  }
  if (node.finalizer) return code + `finally { ${serialize(node.finalizer)} }`;
  return code;
};

export const ForStatement = (node: Types.ForStatement) => {
  return `for (${serialize(node.init)}; ${serialize(node.test)}; ${serialize(
    node.update
  )}) { ${serialize(node.body)} }`;
};

export const ForInStatement = (node: Types.ForInStatement) => {
  return `for (${serialize(node.left)} in ${serialize(
    node.right
  )}) { ${serialize(node.body)} }`;
};

export const ForOfStatement = (node: Types.ForOfStatement) => {
  return `for (${serialize(node.left)} of ${serialize(
    node.right
  )}) { ${serialize(node.body)} }`;
};

export const WhileStatement = (node: Types.WhileStatement) => {
  return `while (${serialize(node.test)}) { ${serialize(node.body)} }`;
};

export const SwitchStatement = (node: Types.SwitchStatement) => {
  return `switch (${serialize(node.discriminant)}) { ${node.cases
    .map(serialize)
    .join("; ")} }`;
};

export const SwitchCase = (node: Types.SwitchCase) => {
  return (
    (node.test ? `case ${serialize(node.test)}: ` : " default: ") +
    node.consequent.map(serialize).join("; ")
  );
};

export const AssignmentExpression = (node: Types.AssignmentExpression) => {
  const left = serialize(node.left);
  const right = serialize(node.right);
  return `${left} ${node.operator} ${right}`;
};

export const BinaryExpression = (node: Types.BinaryExpression) => {
  const left = serialize(node.left);
  const right = serialize(node.right);
  return `(${left} ${node.operator} ${right})`;
};

export const nodes = {
  serialize,
  JsonExpression,
  AssignmentExpression,
  ExpressionStatement,
  Identifier,
  Literal,
  BlockStatement,
  ArrowFunctionExpression,
  FunctionExpression,
  CallExpression,
  MemberExpression,
  ArrayExpression,
  VariableDeclaration,
  VariableDeclarator,
  ReturnStatement,
  IfStatement,
  BreakStatement,
  ContinueStatement,
  ThrowStatement,
  TryStatement,
  ForStatement,
  ForInStatement,
  ForOfStatement,
  WhileStatement,
  SwitchStatement,
  SwitchCase,
  BinaryExpression,
};
