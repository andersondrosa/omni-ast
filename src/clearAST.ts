//
const removeKeys = ["start", "end", "method"];

const exceptions = {
  FunctionDeclaration: (key) => key == "expression",
  FunctionExpression: (key) => key == "expression",
  ArrowFunctionExpression: (key) => key == "id",
};

export function clearAST(ast): any {
  //
  function clear(ast, visited = new WeakMap<object, boolean>()): any {
    //
    if (ast === null) return null;

    if (typeof ast != "object") throw Error("Invalid AST");

    if (visited.has(ast)) throw new Error("Detected a circular AST reference");
    visited.set(ast, true);

    const res = {};

    if (Array.isArray(ast)) return ast.map((x) => clear(x, visited));

    Object.entries(ast).forEach(([key, value]) => {
      //
      if (removeKeys.includes(key)) return;
      if (exceptions[ast.type]?.(key, value)) return;

      if (typeof value != "object" || value instanceof RegExp || value === null)
        return (res[key] = value);

      res[key] = clear(value);
    });

    visited.delete(ast);

    return res;
  }
  return clear(ast);
}
