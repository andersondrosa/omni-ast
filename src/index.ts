import { generate } from "./JsonGenerate";

const isObject = (x) =>
  typeof x === "object" && !Array.isArray(x) && x !== null;

export { generate };
