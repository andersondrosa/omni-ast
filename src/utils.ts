import { ObjectExpressionToJSON } from "./JsonGenerate";
import { Node, Property } from "../src/types";
import { lit, objectExpression } from "./builders";
import { serialize } from "./generators";

type Options = { pre?; post?; skipProperty? };

export function simpleTraverse(root, options: Options = {}) {
  var pre = options.pre;
  var post = options.post;
  var skipProperty = options.skipProperty;

  function visit(node, parent, prop?, idx?) {
    if (!node || typeof node.type !== "string") return;

    var res = undefined;
    if (pre) res = pre(node, parent, prop, idx);

    if (res !== false) {
      for (const prop in node) {
        if (skipProperty ? skipProperty(prop, node) : prop[0] === "$") continue;
        const child = node[prop];
        if (Array.isArray(child)) {
          for (var i = 0; i < child.length; i++) visit(child[i], node, prop, i);
        } else visit(child, node, prop);
      }
    }

    if (post) post(node, parent, prop, idx);
  }

  visit(root, null);
}

export const Mock = new Proxy({ fn: {} } as any, {
  get(target, name, receiver) {
    return { type: "Identifier", name };
  },
});

const isObject = (item) =>
  item && typeof item === "object" && !Array.isArray(item);

export function find(node, match) {
  //
  if (typeof node != "object" || node === null) return;

  const recursive = (object: Object) => {
    //
    if (match(object)) return object;

    if (Array.isArray(object)) {
      for (const val of object) {
        const res = recursive(val);
        if (res) return res;
      }
      return;
    }

    for (const i in object) {
      const value = object[i];
      if (typeof value != "object" || value === null) continue;
      const res = recursive(value);
      if (res) return res;
    }
  };

  return recursive(node);
}

export function mutate(object, match, modifier) {
  const findAndModify = (object) => {
    if (isObject(object)) {
      console.log(typeof match);
      if (match(object)) return modifier(object);
      const newObj = {};
      let changed = false;
      for (const [k, v] of Object.entries(object)) {
        const newValue = findAndModify(v);
        newObj[k] = newValue;
        if (newValue !== v) changed = true;
      }
      return changed ? newObj : object;
    }
    if (Array.isArray(object)) {
      const newArray = object.map(findAndModify);
      return newArray.some((item, index) => item !== object[index])
        ? newArray
        : object;
    }
    return object;
  };
  return findAndModify(object);
}

const denied = ["start", "end", "generator"];
const hideIfNull = ["async", "id", "label", "computed", "optional"];

export function cleanAST(ast): any {
  if (typeof ast != "object" || ast === null) return ast;
  const res = {};
  for (const key in ast) {
    if (denied.includes(key)) continue;
    const value = ast[key];
    if (hideIfNull.includes(key)) {
      // console.log(">>", key, value);
      if (!value) continue;
    }
    if (typeof value != "object") {
      res[key] = value;
      continue;
    }
    if (Array.isArray(value)) {
      res[key] = value.map((x) => cleanAST(x));
      continue;
    }
    res[key] = cleanAST(value);
  }
  return res;
}

export function parseOmniAST(ast: Node) {
  //
  if (typeof ast != "object" || ast === null) return ast;

  if (Array.isArray(ast)) return ast.map(parseOmniAST);

  if (!ast.hasOwnProperty("type")) return ast;

  if (typeof ast["type"] != "string")
    throw Error("Invalid AST. Type is invalid");

  if (ast.type == "ObjectExpression")
    return { type: "JsonExpression", body: ObjectExpressionToJSON(ast) };

  const omniAst = {};
  for (const key in ast) {
    omniAst[key] = parseOmniAST(ast[key]);
  }
  return omniAst;
}

export function parseAST(value: Node) {
  //
  const astFindJSON = (ast: Node) => {
    //
    if (Array.isArray(ast)) return ast.map(astFindJSON);

    if (ast.type == "JsonExpression") return jsonToAST(ast.body);

    const _ast: Node = {} as Node;
    for (const key in ast) {
      _ast[key] =
        typeof ast[key] === "object" && ast[key] != null
          ? astFindJSON(ast[key])
          : ast[key];
    }
    return _ast;
  };

  const jsonToAST = (value) => {
    //
    if (typeof value != "object") return lit(value);

    if (Array.isArray(value))
      return { type: "ArrayExpression", elements: value.map(jsonToAST) };

    if (value.type == "#AST") return astFindJSON(value.body);

    const properties: Property[] = [];

    for (const key in value) {
      const row = value[key];
      properties.push({
        type: "Property",
        key: lit(key),
        value:
          typeof row != "object" || row === null ? lit(row) : jsonToAST(row),
        computed: false,
        kind: "init",
        method: false,
        shorthand: false,
      });
    }
    return objectExpression(properties);
  };

  return astFindJSON(value);
}

export const tryGenerate = (ast) => {
  try {
    return serialize(ast);
  } catch (e) {
    return `[${e.message}]`;
  }
};
