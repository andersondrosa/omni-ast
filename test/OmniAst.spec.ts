import { describe, expect, it } from "vitest";
import { serialize } from "../src/Serialize";
import { tokenizer } from "./utils/tokenizer";
import {
  arrowFunctionExpression,
  ast,
  blockStatement,
  callExpression,
  constantDeclaration,
  functionExpression,
  identifier,
  jsonExpression,
  memberExpression,
  returnStatement,
} from "../src/builders";
import { cleanAST, parseAST, parseOmniAST } from "../src/utils";
import { generate } from "../src";
import { Node, VariableDeclaration } from "../src/types";

const acorn = require("acorn");
const options = { ecmaVersion: "latest" };
const dir = (x) => console.dir(x, { depth: 12 });

describe("OmniAST", () => {
  //
  it("Should Works", () => {
    //
    const script = `fn({ aaa: "bar" })`;

    const AST = cleanAST(acorn.parse(script, options)).body[0].expression;

    // const omniAST = parseOmniAST(AST);
    // dir(omniAST);

    const omniAST = {
      type: "CallExpression",
      callee: { type: "Identifier", name: "fn" },
      arguments: [{ type: "JsonExpression", body: { aaa: "bar" } }],
      optional: false,
    };

    const code = serialize(omniAST);

    // console.log(script, "\n>>");
    // console.log(code);

    expect(tokenizer(script)).toMatchObject(tokenizer(code));
  });

  //

  it("Render test", () => {
    const script = `({ 
      json: "here", 
      myFunc: jsAgain({ json: "again" }) 
    })`;

    const realAST = acorn.parse(script, options).body[0].expression;
    const cleanAcornAST = cleanAST(realAST);

    dir(cleanAcornAST);
    
    const omniAst = parseOmniAST(cleanAcornAST as Node);

    dir(omniAst);

    const generated = `(${serialize(omniAst)})`;

    console.log(generated);
  });

  it("Hybrid", () => {
    const script = `({ 
      json: "here", 
      myFunc: jsAgain({ json: "again" }) 
    })`;

    const realAST = acorn.parse(script, options).body[0].expression;

    expect(realAST).toMatchSnapshot();

    const cleanAcornAST = cleanAST(realAST);

    expect(cleanAcornAST).toMatchSnapshot();

    const omniAst = parseOmniAST(cleanAcornAST as Node);

    const AST = parseAST(omniAst);

    // expect(AST).toMatchObject(cleanAcornAST);

    const generated = `(${serialize(omniAst)})`;

    // console.log(tokenizer(script), "\n>>");
    console.log(tokenizer(generated));

    // expect(tokenizer(script)).toMatchObject(tokenizer(generated));
  });

  it("Test", () => {
    const script = `({ myVar: myFn(['value']) })`;

    const ast = acorn.parse(script, options).body[0].expression;

    const AST = parseOmniAST(cleanAST(ast));

    const generated = `(${serialize(AST)})`;

    expect(tokenizer(script)).toMatchObject(tokenizer(generated));
  });

  it("Test generate", () => {
    //
    const data = ast(
      arrowFunctionExpression(
        [identifier("alpha")],
        blockStatement([
          returnStatement(
            callExpression(identifier("test"), [
              jsonExpression({ "#id": "main", name: "Anderson" }),
              identifier(""),
            ])
          ),
        ])
      )
    );

    console.dir(data, { depth: 10 });
    const generated = generate({ data });

    console.log(generated);
    // const id = (id) => (x) => x["#id"] == id;

    // const _data = mutate(data, id("main"), assoc("name", "Shigeru"));

    // const modified = find(_data, id("main"));

    // log(modified);

    // mutate(data, ["#id", "main"], (state) => {
    //   return state;
    // });

    // const generated = generate({ data });

    // log(generated);
  });
});
