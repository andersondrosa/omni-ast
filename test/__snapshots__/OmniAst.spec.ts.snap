// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OmniAST > Hybrid 1`] = `
Node {
  "end": 72,
  "properties": [
    Node {
      "computed": false,
      "end": 22,
      "key": Node {
        "end": 14,
        "name": "json",
        "start": 10,
        "type": "Identifier",
      },
      "kind": "init",
      "method": false,
      "shorthand": false,
      "start": 10,
      "type": "Property",
      "value": Node {
        "end": 22,
        "raw": "\\"here\\"",
        "start": 16,
        "type": "Literal",
        "value": "here",
      },
    },
    Node {
      "computed": false,
      "end": 65,
      "key": Node {
        "end": 37,
        "name": "myFunc",
        "start": 31,
        "type": "Identifier",
      },
      "kind": "init",
      "method": false,
      "shorthand": false,
      "start": 31,
      "type": "Property",
      "value": Node {
        "arguments": [
          Node {
            "end": 64,
            "properties": [
              Node {
                "computed": false,
                "end": 62,
                "key": Node {
                  "end": 53,
                  "name": "json",
                  "start": 49,
                  "type": "Identifier",
                },
                "kind": "init",
                "method": false,
                "shorthand": false,
                "start": 49,
                "type": "Property",
                "value": Node {
                  "end": 62,
                  "raw": "\\"again\\"",
                  "start": 55,
                  "type": "Literal",
                  "value": "again",
                },
              },
            ],
            "start": 47,
            "type": "ObjectExpression",
          },
        ],
        "callee": Node {
          "end": 46,
          "name": "jsAgain",
          "start": 39,
          "type": "Identifier",
        },
        "end": 65,
        "optional": false,
        "start": 39,
        "type": "CallExpression",
      },
    },
  ],
  "start": 1,
  "type": "ObjectExpression",
}
`;

exports[`OmniAST > Hybrid 2`] = `
{
  "properties": [
    {
      "computed": false,
      "key": {
        "name": "json",
        "type": "Identifier",
      },
      "kind": "init",
      "method": false,
      "shorthand": false,
      "type": "Property",
      "value": {
        "raw": "\\"here\\"",
        "type": "Literal",
        "value": "here",
      },
    },
    {
      "computed": false,
      "key": {
        "name": "myFunc",
        "type": "Identifier",
      },
      "kind": "init",
      "method": false,
      "shorthand": false,
      "type": "Property",
      "value": {
        "arguments": [
          {
            "properties": [
              {
                "computed": false,
                "key": {
                  "name": "json",
                  "type": "Identifier",
                },
                "kind": "init",
                "method": false,
                "shorthand": false,
                "type": "Property",
                "value": {
                  "raw": "\\"again\\"",
                  "type": "Literal",
                  "value": "again",
                },
              },
            ],
            "type": "ObjectExpression",
          },
        ],
        "callee": {
          "name": "jsAgain",
          "type": "Identifier",
        },
        "optional": false,
        "type": "CallExpression",
      },
    },
  ],
  "type": "ObjectExpression",
}
`;
